src/mlpack/methods/ann/activation_functions/softshrink_function.hpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * @file softshrink_function.hpp
+ * @author Lakshya Ojha
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SOFT_SHRINK_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SOFT_SHRINK_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <algorithm>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * SoftShrinkage operator is defined as,
+ *         | x - lambda, if x >  lambda
+ *  f(x) = | x + lambda, if x < -lambda
+ *         | 0, otherwise
+ *
+ * lambda is set to 0.5 by default.
+ *
+ *          | 1 , if x >  lambda
+ *  f'(x) = | 1 , if x < -lambda
+ *          | 0 , otherwise
+ */
+class SoftShrinkFunction
+{
+ public:
+  /**
+   * Computes the Soft Shrinkage function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x, const double lambda = 0.5)
+  {
+    if (x > lambda)
+    {
+      return x - lambda;
+    }
+    else if (x < -1 * lambda)
+    {
+      return x + lambda;
+    }
+    return 0.;
+  }
+
+  /**
+   * Computes the Soft Shrinkage function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template <typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType &x, OutputVecType &y,
+                  const double lambda = 0.5)
+  {
+    y = x;
+    y.transform([](const double x, const double lambda = 0.5){
+                  return Fn(x, lambda);});
+  }
+
+  /**
+   * Computes the first derivative of the Soft Shrinkage function.
+   *
+   * @param y Input data.
+   * @return f'(x)
+   */
+  static double Deriv(const double y, const double lambda = 0.5)
+  {
+    if (y == 0)
+      return 0;
+    return 1;
+  }
```

Comment:
What is the relevance of lambda here, if not using 

Also your information about the derivative's says

```
  *          | 1 , if x >  lambda
  *  f'(x) = | 1 , if x < -lambda
  *          | 0 , otherwise
```

so if that's case then the derivative's function is not apt, Could you recheck once
if y is 0.3 then According to your description it should be 0, but your definition would give 1

Let me know if I have to clear something

Reply:
Hi @jeffin143 regarding the non matching values if I define the function a little bit better it will all get cleared up.
Firstly the derivative function take an input let's call it y and then generates it's derivative y'.
Now, y is a function of x i.e, **y = f(x)**  and **y'** can be written as a function of x or y as
**y' = f'(x) = g(y)**. I have attached a write up for the same kindly refer.
Now in my comment I have explained the derivative in terms of x not in terms of y so when you take the input **y = 0.3** it is merely y and not x.
So the deriv function will take input 0.3 seeing the function of **y' = g(y)** we get 1st case y > 0 therefore answer we get is 1 and not 0.
![New Doc 2020-02-08 12 22 03_1](https://user-images.githubusercontent.com/57477999/74080877-79e30300-4a6e-11ea-841f-0e2d51d33772.jpg)
Although I think I went a little overboard XD.
Let me know if the explanation is correct.
Also, now I feel the need that this should also be shown in the comments of the code where I have defined the f'(x) derivative. Let me know if it is good to do so.

Reply:
That was a good move , Completely missed it , looks good from my side also 

On a first look i just overlooked it :) +1